File: store.go
package main

import (
	"bufio"
	"encoding/json"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// Sample is one captured upstream request/response pair.
type Sample struct {
	Time            time.Time `json:"time"`
	Host            string    `json:"host"`
	Path            string    `json:"path"`
	RequestedModel  string    `json:"requested_model"`  // model field CC Switch sent upstream
	ReportedModel   string    `json:"reported_model"`   // upstream self-reported (response.completed)
	InputTokens     int64     `json:"input_tokens"`     // upstream-reported prompt tokens
	OutputTokens    int64     `json:"output_tokens"`
	ReasoningTokens int64     `json:"reasoning_tokens"`
	StatusCode      int       `json:"status_code"`
	PromptTokensO2  int       `json:"prompt_tokens_o200k"` // our o200k count of the exact sent text
	PromptTokensCl  int       `json:"prompt_tokens_cl100k"`
	TextChars       int       `json:"text_chars"`
}

type Store struct {
	mu       sync.RWMutex
	samples  []Sample
	path     string
	lastSeen time.Time
}

// maxSamples bounds the in-memory sample slice so a long-running guard doesn't grow
// without limit (the on-disk jsonl stays append-only). Kept high enough that browsing
// recent days is unaffected; only the oldest samples are dropped once exceeded. Var,
// not const, so tests can shrink it.
var maxSamples = 200000

// trimLocked drops the oldest samples once the slice exceeds maxSamples. Caller holds mu.
func (s *Store) trimLocked() {
	if len(s.samples) > maxSamples {
		s.samples = append(s.samples[:0], s.samples[len(s.samples)-maxSamples:]...)
	}
}

func (s *Store) LastSeen() time.Time {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.lastSeen
}

func newStore() *Store {
	dir, _ := dataDir()
	return &Store{path: filepath.Join(dir, "samples.jsonl")}
}

func (s *Store) load() error {
	f, err := os.Open(s.path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return err
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	sc.Buffer(make([]byte, 1024*1024), 8*1024*1024)
	for sc.Scan() {
		line := sc.Bytes()
		if len(line) == 0 {
			continue
		}
		var sample Sample
		if json.Unmarshal(line, &sample) == nil {
			s.samples = append(s.samples, sample)
		}
	}
	s.trimLocked()
	return sc.Err()
}

func (s *Store) add(sample Sample) {
	s.mu.Lock()
	s.samples = append(s.samples, sample)
	s.trimLocked()
	s.lastSeen = time.Now()
	s.mu.Unlock()
	// Append-only persistence; best effort.
	if f, err := os.OpenFile(s.path, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0o600); err == nil {
		if b, err := json.Marshal(sample); err == nil {
			f.Write(append(b, '\n'))
		}
		f.Close()
	}
}

// snapshot returns samples within the given local-day [start,end).
func (s *Store) snapshot(start, end time.Time) []Sample {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]Sample, 0, len(s.samples))
	for _, sample := range s.samples {
		if !sample.Time.Before(start) && sample.Time.Before(end) {
			out = append(out, sample)
		}
	}
	return out
}

File: src-tauri/src/store.rs
//! 抓取样本存储：内存 + 追加式 jsonl（数据留在本机）。
use serde::{Deserialize, Serialize};
use std::path::PathBuf;
use std::sync::Mutex;

/// 一条经 MITM 观察到的上游请求/响应配对。
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Sample {
    pub ts: i64,
    pub host: String,
    pub requested_model: String,       // 请求体里的 model
    pub reported_model: String,        // 上游自报（response.completed）
    pub input_tokens: i64,             // 上游上报的 prompt tokens
    pub output_tokens: i64,
    pub reasoning_tokens: i64,
    pub status: u16,
    pub prompt_tokens_o200k: i64,      // 我们对发出正文的 o200k 计数
    pub prompt_tokens_cl100k: i64,
    pub text_chars: i64,
}

pub struct Store {
    samples: Mutex<Vec<Sample>>,
    path: PathBuf,
}

impl Store {
    pub fn new() -> Self {
        let dir = crate::paths::data_dir();
        let path = dir.join("samples.jsonl");
        let mut samples = Vec::new();
        if let Ok(text) = std::fs::read_to_string(&path) {
            for line in text.lines() {
                if line.trim().is_empty() {
                    continue;
                }
                if let Ok(s) = serde_json::from_str::<Sample>(line) {
                    samples.push(s);
                }
            }
        }
        Store { samples: Mutex::new(samples), path }
    }

    pub fn add(&self, sample: Sample) {
        if let Ok(line) = serde_json::to_string(&sample) {
            use std::io::Write;
            if let Ok(mut f) = std::fs::OpenOptions::new().create(true).append(true).open(&self.path) {
                let _ = writeln!(f, "{line}");
            }
        }
        if let Ok(mut v) = self.samples.lock() {
            v.push(sample);
        }
    }

    /// 某天 [start,end) 的样本快照。
    pub fn snapshot(&self, start: i64, end: i64) -> Vec<Sample> {
        self.samples
            .lock()
            .map(|v| v.iter().filter(|s| s.ts >= start && s.ts < end).cloned().collect())
            .unwrap_or_default()
    }
}
